src/deurlname.1
.\"                                      -*- nroff -*-
.\" deurlname.1 - Manual page for deurlname.
.\"
.\" Copyright (C) 2001, 2002, 2004, 2005, 2007, 2008 Oskar Liljeblad
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU Library General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.\"
.TH DEURLNAME "1" "December 4, 2007" "deurlname (renameutils)"
.SH NAME
deurlname \- Remove URL-encoded characters from file names.
.SH SYNOPSIS
.B deurlname
.RI [ OPTION ]... " " [ FILE ]...
.SH DESCRIPTION
This manual page document describes the \fBdeurlname\fP command.

\fBdeurlname\fP will rename every file specified on the command line
that has URL-encoded characters in its name. The URL-encoded
characters will be translated into the corresponding byte values.
That is, %20 will become space, %41 A and so on.
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
.TP
.B \-v, \-\-verbose
Display every rename done.
.TP
.B \-\-help
Show summary of options.
.TP
.B \-\-version
Output version information and exit.
.SH REPORTING BUGS
Report bugs to <\fIoskar@osk.mine.nu\fP>.
.SH AUTHOR
The author of \fBrenameutils\fP and this manual page is Oskar Liljeblad <\fIoskar@osk.mine.nu\fP>.
.SH COPYRIGHT
Copyright \(co 2001, 2002, 2004, 2005, 2007, 2008 Oskar Liljeblad

This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


NEWS
2008-09-21: renameutils 0.10.0 released.
  Change license to GPL v3.
  Added "import" command, suggested by Lou Afonso.
  Added "--command" option to specify command to use when renaming/copying
  files with qcmd, suggested by Nick Pope and Ulrich Dangel.
  Applied manual page lintian warning patch from Francois Marier.

2007-12-04: renameutils 0.9.0 released.
  Clean debian files in distclean-local, not clean-local.
  Do not remove debian/compat in distclean-local.
  Detect mv and cp options properly for icmd.
  Fix handling of trailing slashes in icmd (reported by Francois Marier).
  Use AC_SYS_LARGEFILE to fix renaming problems with large files (Debian
  bug report).
  Use program_name, not program_invocation_name for improved portability
  (bug reported by Andreas Eriksson).
  Various build and compatibility (Gnulib) fixes.
  Change icmd behavior when multiple arguments are specified - add option
  "--pass-though".
  Miscellaneous icmd improvements. Merge imv and icp manual page.
  Fix `plan' command crash and other crashes, reported by Mike Frysinger.
  Updated French and Swedish translations.

2005-08-23: renameutils 0.8.1 released.
  Now with complete French translation.

2005-08-23: renameutils 0.8.0 released.
  Removed debugging statements rendering imv and icp unusable (oops!)
  (reported by Francois Marier).
  Use mv(1) to perform the actual rename. Fixes Debian bug 316113
  (http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=316113). (Reported by
  Daniel Hulme through Francois Marier.)
  Improved some messages and error messages.
  Change order of inclusion of readline headers, fixing complation problem
  on some distributions (patch from Brian Hall).
  Added --command and --arg-options option to icmd (icp and imv).

2005-08-16: web site updated.
  Add some missing entries to this history.

2005-08-16: renameutils 0.7.0 released
  Added Swedish translation.
  Added French translation (from Francois Marier).
  The executables in the RPM packages now links with ncurses to prevent
  problems on Red Hat and Fedora systems.
  Handle POSIXLY_CORRECT in imv (fixes rename bug) (from Brian Hall).
  Change imv to pass command line through to mv if at least two file names
  are specified (or --target-directory option for mv) (suggested by
  Francois Marier).
  imv now checks if the source file exists prior to asking for destination
  name (suggested by Francois Marier).
  imv now detects and handles options requiring arguments for mv.
  imv now trims the edited file name. This means that the destination file
  name cannot have leading or trailing white space.
  Merge code base for imv and icp into icmd.
  Got rid of year ranges as recommended by the FSF.
  Improve the manual pages a little.
  Add Gnulib modules quote, dirname. Remove stat as it no longer exists in
  Gnulib.
  Fixed some typos in messages of the program.

2005-06-28: web site updated.
  Update description in README and web page.

2005-06-28: renameutils 0.6.0 released.
  Added new programs, icp and qcp (suggested by Mario Vanoni).
  Quote file names and strings before printing them (like GNU coreutils).
  Make use of more gnulib modules (quotearg, yesno, version-etc).
  Fixed a serious memory management-related bug in imv causing renaming to
  fail for no reason.
  Fixed a potential segmentation fault when printing a particular error
  message.
  Added copyright and license header to source files and manual pages.
  Manual page updates:
    Added manual page for deurlname.
    Corrected grammar.
    Removed section in imv about placing it in the Midnight Commander menu.
  Gettext is now external (no intl directory).
  Use Gnulib properly (modules.gnulib and autogen.sh removed).
  Added debian package.
  Added MANIFEST.sources file.
  Fix potential bug in cross-reference solution code.
  Added documentation for qcmd --help from Stephen Kolupaev, thanks.

2004-09-17: renameutils 0.5.1 released.
  Clarified README about requirements.
  Include Makefile.am.head in lib and m4 directories in distribution
  tarball.
  Minor clean ups to the autogen.sh, Makefile.am and configure.ac files.
  Now compiles with --disable-nls as well:
    Don't include (<gettext.h is enough).
    Include <locale.h> where setlocale is used.
  Check that locale was set correctly.

2004-08-26: renameutils 0.5.0 released.
  Added deurlname - a program to remove URL-escaped character sequences
  from filenames (such as %20 for space).
  Internationalized renameutils - added support for gettext.
  Removed debian support since renameutils has been in the official Debian
  archives for quite some time.
  Removed show-control-chars variable and option (automaticly assumed since
  we are using gettext).
  Fixed a bug in dual-column editformat that would cause a segfault when
  renaming files with trailing space in their names (reported by Chris
  Jantzen).
  Fixed a bug in qmv that made it impossible to rename files larger than
  2GB (reported by Chris Gentle).
  Corrected makefiles to allow for creation of staticly linked executables.
  renameutils now requires a C89-compatible compiler and C library.
  Modules from gnulib are used to improve compatibility.

2002-08-07: renameutils 0.4.0 released.
  Enable autowidth for dual-column by default, for real this time.
  Fixed "Illegal instruction" crash with '-fdc -oseparate'.
  Added more checks for headers and functions to the configure script.

2002-01-15: renameutils 0.3.0 released.
  Added new option `autowidth' to the dual-column edit format, which is
  enabled by default.
  Handle case when the edited file can't be read correctly (for whatever
  reason).
  Fixed bug in imv that made it impossible to overwrite files.
  Now compiles with more error checking (compile with -Wall).
  Fixed some compilation errors.
  Added files for making an RPM package of renameutils.

2001-10-25: renameutils 0.2.0 released.
  Added manual pages for qmv and imv.
  Added files for making a renameutils Debian package.

src/deurlname.c
/* deurlname.c - Remove URL-encoded characters from file names
 *
 * Copyright (C) 2001, 2002, 2004, 2005, 2007, 2008 Oskar Liljeblad
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <config.h>
#include <stdio.h>  	    	    /* C89 */
#include <string.h> 	    	    /* gnulib (C89) */
#include <stdlib.h> 	    	    /* C89 */
#include <ctype.h>  	    	    /* C89 */
#include <locale.h> 	    	    /* gnulib (POSIX) */
#include <getopt.h> 	    	    /* gnulib (POSIX) */
#include <stdbool.h>	    	    /* gnulib (POSIX) */
#include <gettext.h> 	    	    /* gnulib (gettext) */
#define _(String) gettext(String)
#define N_(String) (String)
#include "progname.h"	    	    /* gnulib */
#include "version-etc.h"	    /* gnulib */
#include "quotearg.h"	    	    /* gnulib */
#include "configmake.h"		    /* gnulib */
#include "common/error.h"
#include "common/strbuf.h"

#define PROGRAM "deurlname"

const char version_etc_copyright[] = 
    "Copyright (C) 2001, 2002, 2004, 2005, 2007, 2008 Oskar Liljeblad";

enum {
    VERSION_OPT = 256,
    HELP_OPT
};

static struct option option_table[] = {
    { "verbose", no_argument, NULL, 'v' },
    { "version", no_argument, NULL, VERSION_OPT },
    { "help",    no_argument, NULL, HELP_OPT    },
    { 0 }
};

static inline int
hexvalue (char c)
{
    if (c >= '0' && c <= '9')
    	return c-'0';
    if (c >= 'A' && c <= 'F')
    	return c-'A'+10;
    return c-'a'+10;
}

int
main (int argc, char **argv)
{
    StrBuf *newnamebuf;
    bool verbose = false;
    int c;
    int rc = EXIT_SUCCESS;

    set_program_name(argv[0]);

    if (setlocale(LC_ALL, "") == NULL)
    	warn(_("cannot set locale: %s\n"), errstr);
#ifdef ENABLE_NLS
    if (bindtextdomain(PACKAGE, LOCALEDIR) == NULL)
        warn(_("cannot bind message domain: %s\n"), errstr);
    if (textdomain(PACKAGE) == NULL)
    	warn(_("cannot set message domain: %s\n"), errstr);
#endif

    while (true) {
    	c = getopt_long (argc, argv, "v", option_table, NULL);
        if (c == -1)
            break;

        switch (c) {
        case '?':
            exit(EXIT_FAILURE);
        case 'v': /* --verbose */
            verbose = true;
            break;
        case VERSION_OPT:
	    version_etc(stdout, PROGRAM, PACKAGE, VERSION, "Oskar Liljeblad", NULL);
  	    exit(EXIT_SUCCESS);
        case HELP_OPT:
	    printf(_("Usage: %s [OPTION]... [FILE]...\n\
Remove URL-encoded characters from file names.\n\
\n\
Options:\n\
  -v, --verbose              display every rename done\n\
      --help                 display this help and exit\n\
      --version              output version information and exit\n\
\n\
The encoded slash character (%%2f) is left untouched in file names.\n"), program_name);
            printf(_("\nReport bugs to <%s>.\n"), PACKAGE_BUGREPORT);
  	    exit(EXIT_SUCCESS);
        }
    }

    if (optind >= argc)
    	die(_("missing file argument"));

    newnamebuf = strbuf_new();
    for (c = optind; c < argc; c++) {
    	char *name;
	char *basename;
	char *newname;
	int d;

    	name = argv[c];
    	strbuf_clear(newnamebuf);
	basename = strrchr(name, '/');
	d = 0;
	if (basename != NULL) {
	    strbuf_append_substring(newnamebuf, name, 0, basename-name+1);
	    d = basename-name+1;
	}
	for (; name[d] != '\0'; d++) {
    	    if (name[d] == '%' && isxdigit(name[d+1]) && isxdigit(name[d+2])) {
	    	char newch = (hexvalue(name[d+1]) << 4) | hexvalue(name[d+2]);
	    	if (newch != '\0' && newch != '/') {
		    strbuf_append_char(newnamebuf, newch);
	    	    d += 2;
	    	    continue;
                }
	    }
	    strbuf_append_char(newnamebuf, name[d]);
	}

        newname = strbuf_buffer(newnamebuf);
	if (strcmp(name, newname) != 0) {
	    if (verbose)
	        printf(_("%s => %s\n"), quotearg_n(0, name), quotearg_n(1, newname));
	    if (rename(name, newname) < 0) {
	        warn(_("cannot rename `%s' to `%s': %s"), quotearg_n(0, name), quotearg_n(1, newname), errstr);
                rc = EXIT_FAILURE;
            }
	}
    }

    strbuf_free(newnamebuf);
    exit(rc);
}
